src/libstd/path.rs

```diff
@@ -0,0 +1,2415 @@
+            self.next()
+        } else { // self.front == State::Suffix
+            self.front = State::Done;
+            if self.prefix_verbatim() {
+                Some(Component::Empty)
+            } else {
+                Some(Component::CurDir)
+            }
+        }
+    }
+}
+
+impl<'a> DoubleEndedIterator for Components<'a> {
+    fn next_back(&mut self) -> Option<Component<'a>> {
+        if self.finished() {
+            None
+        } else if self.back == State::Suffix {
+            self.back = State::Body;
+            if self.prefix_verbatim() {
+                Some(Component::Empty)
+            } else {
+                Some(Component::CurDir)
+            }
+        } else if self.back == State::Body {
+            while self.path.len() > self.prefix_and_root() {
+                let (size, comp) = self.parse_next_component_back();
+                self.path = &self.path[.. self.path.len() - size];
+                if comp.is_some() { return comp }
+            }
+            self.back = State::Root;
+            self.next_back()
+        } else if self.back == State::Root {
+            self.back = State::Prefix;
+            if self.has_physical_root {
+                self.path = &self.path[.. self.path.len() - 1];
+                return Some(Component::RootDir)
+            } else if let Some(p) = self.prefix {
+                if p.has_implicit_root() && !p.is_verbatim() {
+                    return Some(Component::RootDir)
+                }
+            }
+            self.next_back()
+        } else { // self.back == State::Prefix
+            self.back = State::Done;
+            if self.prefix_len() > 0 {
+                //debug_assert!(self.path.len() == self.prefix.len());
+                Some(Component::Prefix(unsafe { Path::from_u8_slice(self.path) }))
+            } else {
+                None
+            }
+        }
+    }
+}
+
+fn optional_path(path: &Path) -> Option<&Path> {
+    if path.as_u8_slice().is_empty() { None } else { Some(path) }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Basic types and traits
+////////////////////////////////////////////////////////////////////////////////
+
+/// An owned, mutable path (akin to `String`).
+pub struct PathBuf {
+    inner: OsString
+}
+
+/// A slice of a path (akin to `str`).
+pub struct Path {
+    inner: OsStr
+}
+
+/// Freely convertible to a `Path`.
+pub trait AsPath {
+    /// Convert to a `Path`.
+    fn as_path(&self) -> &Path;
+}
+
+impl ops::Deref for PathBuf {
+    type Target = Path;
+
+    fn deref(&self) -> &Path {
+        unsafe { mem::transmute(&self.inner[]) }
+    }
+}
+
+impl BorrowFrom<PathBuf> for Path {
+    fn borrow_from(owned: &PathBuf) -> &Path {
+        owned.deref()
+    }
+}
+
+impl AsPath for Path {
+    fn as_path(&self) -> &Path { self }
+}
+
+impl<T: AsOsStr + ?Sized> AsPath for T {
+    fn as_path(&self) -> &Path { Path::from_os_str(self.as_os_str()) }
+}
```

Comment:
Hm, this seems to have worked out interestingly, I have a few questions:

* Shouldn't `Path` implement `AsOsStr`?
* This blanket impl for `T: AsOsStr` does bring into question why `AsPath` is necessary other than just giving it a better name.